bsr_tensor_invariants.md

```diff
@@ -0,0 +1,258 @@
+                       [ 9, 10, 11]],
+
+                      [[12, 13, 14],
+                       [18, 19, 20]],
+
+                      [[15, 16, 17],
+                       [21, 22, 23]]]), size=(4, 6), nnz=4,
+       layout=torch.sparse_bsr)
+```
+Notice that `crow_indices` and `col_indices` contain indices of a CSR
+tensor with shape `(2, 2)` while the shape of BSR tensor is `(4, 6)`
+because it incorporates the shape of `values` blocks: a block is a
+strided tensor with shape `(2, 3)`.
+
+### BSR tensor members
+
+A tensor with BSR layout has the following members (as defined by constructor `sparse_bsr_tensor`):
+- `crow_indices` contains the compressed block row indices information
+- `col_indices` contains block column indices
+- `values` contains the values of tensor elements organized in a 3-dimensional array of blocks
+- `size` defines the shape of tensor
+- `dtype` defines the dtype of tensor elements
+- `layout` holds the layout parameter
+- `device` holds the device of values storage
+- `pin_memory` defines if cuda storage uses pinned memory
+
+### Type invariants
+
+1.1 `crow_indices.dtype == indices_dtype`
+
+1.2 `col_indices.dtype = indices_dtype`
+
+1.3 `indices_dtype` is `int32` (default) or `int64`
+
+1.4 `values.dtype == dtype`
+
+1.5 `dtype` is `float32` (default), `float64`, `int8`, `int16`,
+    `int32`, `int64`, `bool`, `complex32` (unsupported ATM),
+    `complex64`, or `complex128`.
+
+### Layout invariants
+
+2.1 `crow_indices.layout == torch.strided`
+
+2.2 `col_indices.layout == torch.strided`
+
+2.3 `values.layout == torch.strided`
+
+2.4 `layout == torch.sparse_bsr`
+
+### Shape and strides invariants
+
+3.1 `size == (nrows * blocksize[0], ncols * blocksize[1])`, that is, a
+    BSR tensor represents a two dimensional tensor.
+
+3.2 `crow_indices.dim() == 1`
+
+3.3 `col_indices.dim() == 1`
+
+3.4 `values.dim() == 3`
+
+3.5 `crow_indices.is_contiguous()`
+
+3.6 `col_indices.is_contiguous()`
+
+3.7 `values.is_contiguous()` or `values.transpose(-2, -1).is_contiguous()`
+
+3.8 `crow_indices.size() == (nrows + 1,)`
+
+3.9 `col_indices.size() == (nnz,)`
+
+3.10 `values.size() == (nnz, blocksize[0], blocksize[1])`
+
+3.11 `numel() == nrows * ncols * blocksize[0] * blocksize[1]` is the number of indexable elements (note: indexing CSC, BSR, and BSC tensors is not yet supported)
+
+### Device invariants
+
+4.1 `device` is `CPU` or `CUDA`
+
+4.2 `crow_indices.device == device`
+
+4.3 `col_indices.device == device`
+
+4.4 `values.device == device`
+
+### Indices invariants
+
+5.1 `crow_indices[0] == 0`
+
+5.2 `crow_indices[nrows] == nnz`
+
+5.3 `0 <= crow_indices[i] - crow_indices[i-1] <= ncols` for all `i=1,...,nrows`
+
+5.4 `0 <= col_indices.min()`
+
+5.5 `col_indices.max() < ncols`
+
+5.6 `col_indices[crow_indinces[i-1]:crow_indinces[i]]` must be sorted and with distinct values for all `i=1,...,nrows` ([required by cuSparse](https://docs.nvidia.com/cuda/cusparse/index.html#csr-format))
```

Comment:
Just for reference scipy seems to call this condition the [canonical form](https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/sparse/sparsetools/csr.h#L313-L324). I think we can call it "coalesced" in reference to our COO implementation. 

Indeed at some point in the future we might decide to allow uncoalesced sparse Tensors with compressed dimensions, but just like any other performance optimization we should careful weigh pros and cons and collect traces where having such a feature is an advantage.